dnsmasq_interface.c

```diff
@@ -1200,77 +1200,72 @@ static void block_single_domain_regex(char *domain)
 	add_blocked_domain(&addr4, &addr6, has_IPv4, has_IPv6, domain, strlen(domain), NULL, 0, SRC_REGEX);
 
 	if(config.debug & DEBUG_QUERIES) logg("Added %s to cache", domain);
-
-	return;
 }
 
 int FTL_listsfile(char* filename, unsigned int index, FILE *f, int cache_size, struct crec **rhash, int hashsz)
 {
 	int name_count = cache_size;
 	int added = 0;
-	size_t size = 0;
-	char *buffer = NULL;
 	struct all_addr addr4 = {{{ 0 }}}, addr6 = {{{ 0 }}};
 	bool has_IPv4 = false, has_IPv6 = false;
+	unsigned char list = 0;
 
 	// Handle only gravity.list and black.list
 	// Skip all other files (they are interpreted in the usual format)
-	if(strcmp(filename, files.gravity) != 0 &&
-	   strcmp(filename, files.blacklist) != 0)
+	if(strcmp(filename, files.gravity) == 0)
+		list = GRAVITY_LIST;
+	else if(strcmp(filename, files.blacklist) == 0)
+		list = BLACK_LIST;
+	else
+		return cache_size;
+
+	// Jump to end of file to ensure dnsmasq does not
+	// try to read whatever might be in the file on
+	// disk when we return
+	fseek(f, 0, SEEK_END);
+
+	if(blockingstatus == BLOCKING_DISABLED)
+	{
+		logg("Skipping %s because blocking is disabled", filename);
 		return cache_size;
+	}
+	else
+	{
+		logg("Importing content of template %s from database", filename);
```

Comment:
What does "template" mean in this context? If it is just a placeholder "flag" to tell FTL to import from the database, then this message seems to suggest that we're reading from the template instead of the database.

Reply:
We need the domain mass-injection machinery that is prepared by `dnsmasq` for when it reads in HOSTS files. Here, we hijack this method and decide that, if the filenames are `/etc/pihole/gravity.list` or `/etc/pihole/black.list` (or whatever the user might have set manually in `pihole-FTL.conf), we ignore the real content of these files and instead read from the database.

This allows us to have the required tools available while still reading efficiently from the database. My fist implementation tried to manually prepare a suitable artificial mass-injection machinery but was very slow as I'd have had to alter too many cache routines in `dnsmasq`. This way of doing it prevents us from having to make *any* changes to the `dnsmasq` code.

Having this in mind, please suggest an alternative message.

Reply:
My question was just on the word "template" and the message (everything else seems fine). I can't think of a better name now, but I just wanted to understand the reasoning behind the name.